src/hotspot/share/cds/archiveHeapWriter.cpp

```diff
@@ -0,0 +1,658 @@
+void ArchiveHeapWriter::write(GrowableArrayCHeap<oop, mtClassShared>* roots,
+                              GrowableArray<MemRegion>* closed_regions, GrowableArray<MemRegion>* open_regions,
+                              GrowableArray<ArchiveHeapBitmapInfo>* closed_bitmaps,
+                              GrowableArray<ArchiveHeapBitmapInfo>* open_bitmaps) {
+  assert(HeapShared::can_write(), "sanity");
+  allocate_buffer();
+  copy_source_objs_to_buffer(roots);
+  set_requested_address_for_regions(closed_regions, open_regions);
+  relocate_embedded_oops(roots, closed_bitmaps, open_bitmaps);
+}
+
+bool ArchiveHeapWriter::is_too_large_to_archive(oop o) {
+  return is_too_large_to_archive(o->size());
+}
+
+bool ArchiveHeapWriter::is_string_too_large_to_archive(oop string) {
+  typeArrayOop value = java_lang_String::value_no_keepalive(string);
+  return is_too_large_to_archive(value);
+}
+
+bool ArchiveHeapWriter::is_too_large_to_archive(size_t size) {
+  assert(size > 0, "no zero-size object");
+  assert(size * HeapWordSize > size, "no overflow");
+  static_assert(MIN_GC_REGION_ALIGNMENT > 0, "must be positive");
+
+  size_t byte_size = size * HeapWordSize;
+  if (byte_size > size_t(MIN_GC_REGION_ALIGNMENT)) {
+    return true;
+  } else {
+    return false;
+  }
+}
+
+// Various lookup functions between source_obj, buffered_obj and requested_obj
+bool ArchiveHeapWriter::is_in_requested_regions(oop o) {
+  assert(_requested_open_region_bottom != nullptr, "do not call before this is initialized");
+  assert(_requested_closed_region_bottom != nullptr, "do not call before this is initialized");
+
+  address a = cast_from_oop<address>(o);
+  return (_requested_open_region_bottom <= a && a < _requested_open_region_top) ||
+         (_requested_closed_region_bottom <= a && a < _requested_closed_region_top);
+}
+
+oop ArchiveHeapWriter::requested_obj_from_buffer_offset(size_t offset) {
+  oop req_obj = cast_to_oop(_requested_open_region_bottom + offset);
+  assert(is_in_requested_regions(req_obj), "must be");
+  return req_obj;
+}
+
+oop ArchiveHeapWriter::source_obj_to_requested_obj(oop src_obj) {
+  assert(DumpSharedSpaces, "dump-time only");
+  HeapShared::CachedOopInfo* p = HeapShared::archived_object_cache()->get(src_obj);
+  if (p != nullptr) {
+    return requested_obj_from_buffer_offset(p->buffer_offset());
+  } else {
+    return nullptr;
+  }
+}
+
+oop ArchiveHeapWriter::buffered_addr_to_source_obj(address buffered_addr) {
+  oop* p = _buffer_offset_to_source_obj_table->get(buffered_address_to_offset(buffered_addr));
+  if (p != nullptr) {
+    return *p;
+  } else {
+    return nullptr;
+  }
+}
+
+address ArchiveHeapWriter::buffered_addr_to_requested_addr(address buffered_addr) {
+  return _requested_open_region_bottom + buffered_address_to_offset(buffered_addr);
+}
+
+oop ArchiveHeapWriter::heap_roots_requested_address() {
+  return cast_to_oop(_requested_open_region_bottom + _heap_roots_bottom);
```

Comment:
A suggestion: it is probably better to use the internal method `requested_obj_from_buffer_offset(_heap_roots_bottom)` instead.